README.md
# Home Assistant Virtual Nuki Bridge

Use this instead of the [Nuki Bridge](https://nuki.io/en/bridge/) device.

> Important - if you are experiencing delays using RPI, it is advised to use a bluetooth dongle instead of the builtin bluetooth hardware.
> [TP-LINK UB400](https://www.tp-link.com/us/home-networking/usb-adapter/ub400/) is verified to be working.

## Background

- This is heavily inspired by the work of [Jandebeule](https://github.com/jandebeule/nukiPyBridge).
- Forked from [dauden1184](https://github.com/dauden1184/) 's [RaspiNukiBridge](https://github.com/dauden1184/RaspiNukiBridge)
- Similar project [here](https://github.com/ftarolli/NukiBridgeAddon)
- Nuki documentation
  - [Nuki Smart Lock API V2.2.1](https://developer.nuki.io/page/nuki-smart-lock-api-2/2/#heading--lock-action)
  - [Nuki Bridge HTTP API V1.13.2](https://developer.nuki.io/page/nuki-bridge-http-api-1-13/4/#heading--lockstate)

# Installation

There are 2 methods:

1. [As a Home Assistant Addon](#installation-as-home-assistant-os)
   1. Use this if your Home Assistant is nearby the Nuki Smart Lock
2. [Outside of Home Assistant](#installation-outside-of-home-assistant-os)
   1. Use this if your Home Assistant too far from the Nuki Smart Lock

# Installation outside of Home Assistant OS

## Get the files

> ### Raspberry Pi 3B+ and 4 only
>
> DOWNGRADE Bluez. [See comment](https://github.com/dauden1184/RaspiNukiBridge/issues/1#issuecomment-1103969957).
>
> ```
> wget http://ftp.hk.debian.org/debian/pool/main/b/bluez/bluez_5.50-1.2~deb10u2_armhf.deb
> sudo apt install ./bluez_5.50-1.2~deb10u2_armhf.deb
> ```
>
> Reboot the Raspberry Pi

1. Clone the repository.

   ```
   git clone https://github.com/dauden1184/RaspiNukiBridge.git
   ```

2. Go to the addon directoy.

   ```
   cd nuki_bridge
   ```

3. Install the requirements.

   ```
   pip install -r requirements.txt
   ```

4. Go to [pairing](#pairing)

## Start automatically at boot

Create a new systemd service:

```
sudo nano /etc/systemd/system/nukibridge.service
```

Put this content in the file (change 'user' and 'WorkingDirectory' to your needs):

```
[Unit]
Description=Nuki bridge
After=network-online.target
[Service]
Type=simple
Restart=always
RestartSec=1
User=pi
WorkingDirectory=/home/pi/RaspiNukiBridge/nuki_bridge
ExecStart=python .
[Install]
WantedBy=multi-user.target
```

Enable the service and start it:

```
sudo systemctl daemon-reload
sudo systemctl enable nukibridge.service
sudo systemctl start nukibridge.service
```

# Installation as Home Assistant Addon

## Add custom repository

[![Add repository on my Home Assistant][repository-badge]][repository-url]

If you want to do add the repository manually, please follow the procedure highlighted in the [Home Assistant website](https://home-assistant.io/hassio/installing_third_party_addons). Use the following URL to add this repository: https://github.com/regevbr/RaspiNukiBridge

## Install addon

1. Go to Home Assistant -> Configuration -> Add-ons, backups & Supervisor -> add-on store (in the bottom right corner)

   [![Open your Home Assistant instance and show the Supervisor add-on store.](https://my.home-assistant.io/badges/supervisor_store.svg)](https://my.home-assistant.io/redirect/supervisor_store/)

2. On the top right overflow menu ⋮ click the "Check for updates" button
3. Install the addon. It takes a few minutes.
4. _Don't start it yet!_

# Pairing

1. Press the button of the Nuki Smart Lock for 6 seconds. It should light up.
2. Run

   1. If installed as addon, click `start`
   2. If installed outside HA, run

      `python .`

3. Look in logs for:

   ```
   ********************************************************************
   *                                                                  *
   *                         Pairing completed!                       *
   *                            Access Token                          *
   * abcabcabcabcabcabcabcabcabcabcabcabcabcabcabcabcabcabcabcabcabca *
   *                                                                  *
   ********************************************************************
   ```

4. Copy the `Access Token` for a following step.

   In case you missed it, restart the service and check again.

5. Before moving on, you might need to restart the addon. First run is less stable.

# Nuki Addon

1. Install either
   1. [Official Home Assistant Nuki integration](https://www.home-assistant.io/integrations/nuki/)
   2. [hass_nuki_ng](https://github.com/kvj/hass_nuki_ng)
2. Configure the Nuki Addon:
   1. Paste your token from the log above
   2. Address is `127.0.0.1`

# Enjoy!

[repository-badge]: https://img.shields.io/badge/Add%20repository%20to%20my-Home%20Assistant-41BDF5?logo=home-assistant&style=for-the-badge
[repository-url]: https://my.home-assistant.io/redirect/supervisor_add_addon_repository/?repository_url=https%3A%2F%2Fgithub.com%2Fregevbr%2FRaspiNukiBridge


nuki_bridge/CHANGELOG.md
# Changelog

## [0.0.23] - 2022-07-18

### Fixed

- fix #2 - pairing didn't work
- changed CI to work faster in the docker build steps

## [0.0.22] - 2022-06-29

### Fixed

- fix stuck on infinite loop when device is not ready

## [0.0.21] - 2022-06-28

### Fixed

- removed redundant exception tracing while timeout in connection attempt

## [0.0.20] - 2022-06-28

### Fixed

- fixed event loop collisions

## [0.0.19] - 2022-06-28

### Fixed

- fixed unhandled exception

## [0.0.18] - 2022-06-28

### Fixed

- fixed usage of reference before assignment

## [0.0.17] - 2022-06-28

### Fixed

- fixed infinite recursion

## [0.0.16] - 2022-06-28

### Fixed

- fixed no even loop exists exception

## [0.0.15] - 2022-06-28

### Added

- added a more safe way to handle connections

## [0.0.14] - 2022-06-28

### Fixed

- fixed addon configuration parsing

## [0.0.13] - 2022-06-28

### Fixed

- fixed bad yaml configuration

## [0.0.12] - 2022-06-28

### Added

- added ability to configure parameters from addon

## [0.0.11] - 2022-06-27

### Fixed

- fixed NPE

## [0.0.10] - 2022-06-27

### Fixed

- fix hanging stop scan command

## [0.0.9] - 2022-06-27

### Fixed

- fix undesired disconnects

## [0.0.8] - 2022-06-27

### Fixed

- reverted unneeded new behaviour

## [0.0.7] - 2022-06-27

### Fixed

- timeout not passing properly

## [0.0.6] - 2022-06-27

### Added

- timeout optimizations

## [0.0.5] - 2022-06-27

### Fixed

- Don't wait for stop scan before connecting

## [0.0.4] - 2022-06-27

### Fixed

- reference prebuilt images

## [0.0.3] - 2022-06-27

### Added

- alpha version

[0.0.23-alpha.1]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.22...v0.0.23
[0.0.22]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.21...v0.0.22
[0.0.21]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.20...v0.0.21
[0.0.20]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.19...v0.0.20
[0.0.19]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.18...v0.0.19
[0.0.18]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.17...v0.0.18
[0.0.17]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.16...v0.0.17
[0.0.16]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.15...v0.0.16
[0.0.15]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.14...v0.0.15
[0.0.14]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.13...v0.0.14
[0.0.13]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.12...v0.0.13
[0.0.12]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.11...v0.0.12
[0.0.11]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.10...v0.0.11
[0.0.10]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.9...v0.0.10
[0.0.9]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.8...v0.0.9
[0.0.8]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.7...v0.0.8
[0.0.7]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.6...v0.0.7
[0.0.6]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.5...v0.0.6
[0.0.5]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.4...v0.0.5
[0.0.4]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.3...v0.0.4
[0.0.3]: https://github.com/regevbr/RaspiNukiBridge/compare/v0.0.2...v0.0.3


nuki_bridge/config.py
import asyncio
import os
import random

import json
import yaml
from nacl.public import PrivateKey

from consts import CONF_FILE_NAME, ADDON_CONF_FILE_NAME, DATA_PATH
from nuki import NukiManager, Nuki
from scan_ble import find_ble_device
from utils import logger


def get_config_file():
    if os.path.isdir(DATA_PATH):
        return os.path.join(DATA_PATH, CONF_FILE_NAME)
    return CONF_FILE_NAME

def get_addon_config_file():
    if os.path.isdir(DATA_PATH):
        return os.path.join(DATA_PATH, ADDON_CONF_FILE_NAME)
    return ADDON_CONF_FILE_NAME


def init_config(config_file, addon_config_file):
    if os.path.isfile(config_file):
        with open(config_file) as f:
            data = yaml.load(f, Loader=yaml.FullLoader)
            token = data["server"]["token"]
    else:
        app_id, token = _random_app_id_and_token()
        data = {
            'server': {
                'host': '0.0.0.0',
                'port': '8080',
                'adapter': 'hci0',
                'name': 'RaspiNukiBridge',
                'app_id': app_id,
                'token': token
            }
        }
    if os.path.isfile(addon_config_file):
        with open(addon_config_file) as f:
            addon_data = json.load(f)
    else:
        addon_data = {}
    global global_retry
    global global_connection_timeout
    global global_command_timeout
    if "adapter" in addon_data:
        data["server"]["adapter"] = addon_data["adapter"]
    if "retry" in addon_data:
        global_retry = addon_data["retry"]
    if "connection_timeout" in addon_data:
        global_connection_timeout = addon_data["connection_timeout"]
    if "command_timeout" in addon_data:
       global_command_timeout = addon_data["command_timeout"]
    name = data["server"]["name"]
    app_id = data["server"]["app_id"]
    bt_adapter = data["server"].get("adapter", "hci0")
    nuki_manager = NukiManager(name, app_id, bt_adapter)

    if 'smartlock' in data:
        for smartlock in data['smartlock']:
            if global_retry:
                smartlock["retry"] = global_retry
            if global_connection_timeout:
                smartlock["connection_timeout"] = global_connection_timeout
            if global_command_timeout:
                smartlock["command_timeout"] = global_command_timeout
        logger.info(f"********************************************************************")
        logger.info(f"*                                                                  *")
        logger.info(f"*                            Access Token                          *")
        logger.info(f"* {data['server']['token']} *")
        logger.info(f"*                                                                  *")
        logger.info(f"********************************************************************")
        return nuki_manager, data

    # Bridge keys
    bridge_public_key, bridge_private_key = _generate_bridge_keys()
    smartlock = {
        'bridge_public_key': bridge_public_key.hex(),
        'bridge_private_key': bridge_private_key.hex(),
        'connection_timeout': 10,
        'command_timeout': 30,
        'retry': 5
    }
    if global_retry:
        smartlock["retry"] = global_retry
    if global_connection_timeout:
        smartlock["connection_timeout"] = global_connection_timeout
    if global_command_timeout:
        smartlock["command_timeout"] = global_command_timeout

    # Device MAC Address
    nuki_devices = find_ble_device('Nuki_.*', logger)
    if len(nuki_devices) > 1:
        for device in nuki_devices:
            logger.info(device)
        raise ValueError('Multiple Nuki devices found. Please use --pair [MAC_ADDRESS]')
    if not nuki_devices:
        raise ValueError('No Nuki device found')
    address = nuki_devices[0].address

    smartlock['address'] = address

    # Pair
    nuki = Nuki(address, None, None, bridge_public_key, bridge_private_key)
    nuki_manager.add_nuki(nuki)

    loop = asyncio.new_event_loop()

    nuki_manager.start(loop)

    def pairing_completed(paired_nuki):
        nuki_public_key = paired_nuki.nuki_public_key.hex()
        auth_id = paired_nuki.auth_id.hex()
        logger.info(f"Pairing completed, auth_id: {auth_id}")
        logger.info(f"nuki_public_key: {nuki_public_key}")
        logger.info(f"********************************************************************")
        logger.info(f"*                                                                  *")
        logger.info(f"*                         Pairing completed!                       *")
        logger.info(f"*                            Access Token                          *")
        logger.info(f"* {token} *")
        logger.info(f"*                                                                  *")
        logger.info(f"********************************************************************")
        smartlock['nuki_public_key'] = nuki_public_key
        smartlock['auth_id'] = auth_id

        data['smartlock'] = [smartlock]
        yaml.dump(data, open(config_file, 'w'))
        loop.stop()

    loop.create_task(nuki.pair(pairing_completed))
    loop.run_forever()

    return nuki_manager, data


def _random_app_id_and_token():
    app_id = random.getrandbits(32)
    token = random.getrandbits(256).to_bytes(32, "little").hex()
    logger.info(f'Access Token: {token}')
    logger.info(f'app_id: {app_id}')
    return app_id, token


def _generate_bridge_keys():
    logger.info(f"Generating new keys")
    keypair = PrivateKey.generate()
    bridge_public_key = keypair.public_key.__bytes__()
    bridge_private_key = keypair.__bytes__()
    logger.info(f"bridge_public_key: {bridge_public_key.hex()}")
    logger.info(f"bridge_private_key: {bridge_private_key.hex()}")
    return bridge_public_key, bridge_private_key


nuki_bridge/web_server.py
import datetime
import hashlib
import json
import uuid

from aiohttp import web, ClientSession

from utils import logger
from nuki import DeviceType, BridgeType


class WebServer:

    def __init__(self, host, port, token, nuki_manager):
        self._host = host
        self._port = port
        self._token = token
        self.nuki_manager = nuki_manager
        self._start_datetime = None
        self._server_id = uuid.getnode() & 0xFFFFFFFF  # Truncate server_id to 32 bit, OpenHub doesn't like it too big
        self._http_callbacks = [None, None, None]  # Nuki Bridge support up to 3 callbacks

    def start(self):
        app = web.Application()
        app.add_routes([web.get('/info', self.nuki_info),
                        web.get('/list', self.nuki_list),
                        web.get('/lock', self.nuki_lock),
                        web.get('/unlock', self.nuki_unlock),
                        web.get('/lockAction', self.nuki_lockaction),
                        web.get('/lockState', self.nuki_state),
                        web.get('/callback/add', self.callback_add),
                        web.get('/callback/list', self.callback_list),
                        web.get('/callback/remove', self.callback_remove)])
        app.on_startup.append(self._startup)
        web.run_app(app, host=self._host, port=self._port)

    @staticmethod
    def _get_nuki_last_state(nuki):
        state = {"mode": nuki.last_state["nuki_state"].value,
                 "state": nuki.last_state["lock_state"].value,
                 "stateName": nuki.last_state["lock_state"].name,
                 "batteryCritical": nuki.is_battery_critical,
                 "batteryCharging": nuki.is_battery_charging,
                 "batteryChargeState": nuki.battery_percentage,
                 "keypadBatteryCritical": False,  # How to get this from bt api?
                 "doorsensorState": nuki.last_state["door_sensor_state"].value,
                 "doorsensorStateName": nuki.last_state["door_sensor_state"].name,
                 "ringactionTimestamp": None,  # How to get this from bt api?
                 "ringactionState": None,  # How to get this from bt api?
                 "timestamp": nuki.last_state["current_time"].isoformat().split(".")[0],
                 "success": True,
                 }

        if nuki.device_type == DeviceType.OPENER:
            state["ringactionTimestamp"] = nuki.last_state["current_time"].isoformat().split(".")[0]
            state["ringactionState"] = nuki.last_state["last_lock_action_completion_status"]

        return state

    async def _newstate(self, nuki):
        logger.info(f"Nuki new state: {nuki.last_state}")
        if any(self._http_callbacks):
            async with ClientSession() as session:
                for url in filter(None, self._http_callbacks):
                    try:
                        data = {"nukiId": hex(nuki.config["id"])[2:],
                                "deviceType": nuki.device_type.value}  # How to get this from bt api?
                        data.update(self._get_nuki_last_state(nuki))
                        async with session.post(url, data=json.dumps(data)) as resp:
                            await resp.text()
                    except:
                        logger.exception(f"Error on http callbak {url}")

    async def _startup(self, _app):
        self._start_datetime = datetime.datetime.now()
        await self.nuki_manager.start()

    async def callback_add(self, request):
        if not self._check_token(request):
            raise web.HTTPForbidden()
        callback_url = request.query["url"]
        for i, call in enumerate(self._http_callbacks):
            if not call:
                self._http_callbacks[i] = callback_url
                break
        if not self.nuki_manager.newstate_callback:
            self.nuki_manager.newstate_callback = self._newstate
        logger.info(f"Add http callback: {callback_url}")
        return web.Response(text=json.dumps({"success": True}))

    async def callback_list(self, request):
        if not self._check_token(request):
            raise web.HTTPForbidden()
        resp = {"callbacks": [{"id": url_id, "url": url} for url_id, url in enumerate(self._http_callbacks) if url]}
        return web.Response(text=json.dumps(resp))

    async def callback_remove(self, request):
        if not self._check_token(request):
            raise web.HTTPForbidden()
        url_id = request.query["id"]
        self._http_callbacks[int(url_id)] = None
        return web.Response(text=json.dumps({"success": True}))

    async def nuki_list(self, request):
        if not self._check_token(request):
            raise web.HTTPForbidden()
        resp = [{"nukiId": hex(nuki.config["id"])[2:],
                 "deviceType": nuki.device_type.value,  # How to get this from bt api?
                 "name": nuki.config["name"],
                 "lastKnownState": self._get_nuki_last_state(nuki)} for nuki in self.nuki_manager if nuki.config]
        logger.info(f'Listed {len(resp)} devices')
        return web.Response(text=json.dumps(resp))

    async def nuki_info(self, request):
        if not self._check_token(request):
            raise web.HTTPForbidden()
        resp = {"bridgeType": BridgeType.SW.value,
                # The hardwareId should not be sent if bridgeType is BRIDGE_SW, but the homeassistant
                # integration expects it
                "ids": {"hardwareId": self._server_id, "serverId": self._server_id},
                "versions": {"appVersion": "0.1.0"},
                "uptime": (datetime.datetime.now() - self._start_datetime).seconds,
                "currentTime": datetime.datetime.now().isoformat()[:-7] + "Z",
                "serverConnected": False,
                "scanResults": [{"nukiId": hex(nuki.config["id"])[2:],
                                 "type": nuki.device_type.value,  # How to get this from bt api?
                                 "name": nuki.config["name"],
                                 "rssi": nuki.rssi,
                                 "paired": True} for nuki in self.nuki_manager if nuki.config]}
        return web.Response(text=json.dumps(resp))

    def _check_token(self, request):
        token_valid = False
        if "hash" in request.query:
            rnr = request.query["rnr"]
            ts = request.query["ts"]
            hash_256 = hashlib.sha256(f"{ts},{rnr},{self._token}".encode("utf-8")).hexdigest()
            token_valid = hash_256 == request.query["hash"]
        elif "token" in request.query:
            token_valid = self._token == request.query["token"]
        if not token_valid:
            logger.error('Invalid token. Please change token.')
        return token_valid

    async def nuki_lockaction(self, request):
        if not self._check_token(request):
            raise web.HTTPForbidden()
        action = int(request.query["action"])
        n = self.nuki_manager.nuki_by_id(int(request.query["nukiId"], base=16))
        await n.lock_action(action)
        res = json.dumps({"success": True, "batteryCritical": n.is_battery_critical})
        return web.Response(text=res)

    async def nuki_state(self, request):
        if not self._check_token(request):
            raise web.HTTPForbidden()
        n = self.nuki_manager.nuki_by_id(int(request.query["nukiId"], base=16))
        return web.Response(text=json.dumps(self._get_nuki_last_state(n)))

    async def nuki_lock(self, request):
        if not self._check_token(request):
            raise web.HTTPForbidden()
        n = self.nuki_manager.nuki_by_id(int(request.query["nukiId"], base=16))
        await n.lock()
        res = json.dumps({"success": True, "batteryCritical": n.is_battery_critical})
        return web.Response(text=res)

    async def nuki_unlock(self, request):
        if not self._check_token(request):
            raise web.HTTPForbidden()
        n = self.nuki_manager.nuki_by_id(int(request.query["nukiId"], base=16))
        await n.unlock()
        res = json.dumps({"success": True, "batteryCritical": n.is_battery_critical})
        return web.Response(text=res)

nuki_bridge/__main__.py
import asyncio
import logging
import argparse

from config import init_config, _random_app_id_and_token, _generate_bridge_keys, get_config_file, get_addon_config_file
from nuki import Nuki
from scan_ble import find_ble_device
from utils import logger, handler
from web_server import WebServer

logging.getLogger("aiohttp").addHandler(handler)
logging.getLogger("aiohttp").setLevel(logging.ERROR)
logging.getLogger("bleak").addHandler(handler)
logging.getLogger("bleak").setLevel(logging.ERROR)


def _add_devices_to_manager(data, nuki_manager):
    for ls in data["smartlock"]:
        address = ls["address"]
        auth_id = bytes.fromhex(ls["auth_id"])
        nuki_public_key = bytes.fromhex(ls["nuki_public_key"])
        bridge_public_key = bytes.fromhex(ls["bridge_public_key"])
        bridge_private_key = bytes.fromhex(ls["bridge_private_key"])
        n = Nuki(address, auth_id, nuki_public_key, bridge_public_key, bridge_private_key)
        n.retry = ls.get("retry", 5)
        n.connection_timeout = ls.get("connection_timeout", 10)
        n.command_timeout = ls.get("command_timeout", 30)
        nuki_manager.add_nuki(n)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--config", metavar=('CONFIG_FILE',), help="Specify the yaml file to use")
    parser.add_argument("--pair", metavar=('MAC_ADDRESS',), help="Pair to a nuki smartlock")
    parser.add_argument("--generate-config", action='store_true', help="Generate a new configuration file")
    parser.add_argument("--unlock", action='store_true', help="Unlock")
    parser.add_argument("--lock", action='store_true', help="Lock")
    parser.add_argument("--verbose", nargs='?', const=1, type=int, default=0, help="More logs")
    args = parser.parse_args()

    if not args.verbose:
        logger.setLevel(level=logging.INFO)
        logging.getLogger("aiohttp").setLevel(level=logging.ERROR)
        logging.getLogger("bleak").setLevel(level=logging.ERROR)
    elif args.verbose == 1:
        logger.setLevel(level=logging.DEBUG)
        logging.getLogger("aiohttp").setLevel(level=logging.INFO)
        logging.getLogger("bleak").setLevel(level=logging.INFO)
    elif args.verbose == 2:
        logger.setLevel(level=logging.DEBUG)
        logging.getLogger("aiohttp").setLevel(level=logging.DEBUG)
        logging.getLogger("bleak").setLevel(level=logging.DEBUG)

    if args.generate_config:
        app_id, token = _random_app_id_and_token()
        print(f"server:\n"
              f"  host: 0.0.0.0\n"
              f"  port: 8080\n"
              f"  name: RaspiNukiBridge\n"
              f"  app_id: {app_id}\n"
              f"  token: {token}\n")
        exit(0)

    config_file = args.config or get_config_file()
    addon_config_file = get_addon_config_file()
    nuki_manager, data = init_config(config_file, addon_config_file)

    if args.pair:
        if args.pair:
            address = args.pair
        else:
            address = find_ble_device('Nuki_.*', logger)

        bridge_public_key, bridge_private_key = _generate_bridge_keys()
        nuki = Nuki(address, None, None, bridge_public_key, bridge_private_key)
        nuki_manager.add_nuki(nuki)

        loop = asyncio.get_event_loop()

        def pairing_completed(paired_nuki):
            logger.info(f"Pairing completed, nuki_public_key: {paired_nuki.nuki_public_key.hex()}")
            logger.info(f"Pairing completed, auth_id: {paired_nuki.auth_id.hex()}")
            loop.stop()
        loop.create_task(nuki.pair(pairing_completed))
        loop.run_forever()
    else:
        if not nuki_manager.device_list:
            _add_devices_to_manager(data, nuki_manager)

        if args.unlock:
            device = nuki_manager.device_list[0]
            asyncio.run(device.unlock())
        elif args.lock:
            device = nuki_manager.device_list[0]
            asyncio.run(device.lock())
        else:
            host = data["server"]["host"]
            port = data["server"]["port"]
            token = data["server"]["token"]
            web_server = WebServer(host, port, token, nuki_manager)
            web_server.start()
